src/3d/terrain/qgsterraintexturegenerator_p.cpp

```diff
@@ -0,0 +1,122 @@
+#include "qgsterraintexturegenerator_p.h"
+
+#include <qgsmaprenderercustompainterjob.h>
+#include <qgsmaprenderersequentialjob.h>
+#include <qgsmapsettings.h>
+#include <qgsproject.h>
+
+#include "qgs3dmapsettings.h"
+
+///@cond PRIVATE
+
+QgsTerrainTextureGenerator::QgsTerrainTextureGenerator( const Qgs3DMapSettings &map )
+  : mMap( map )
+  , mLastJobId( 0 )
+{
+}
+
+int QgsTerrainTextureGenerator::render( const QgsRectangle &extent, const QString &debugText )
+{
+  QgsMapSettings mapSettings( baseMapSettings() );
+  mapSettings.setExtent( extent );
+
+  QgsMapRendererSequentialJob *job = new QgsMapRendererSequentialJob( mapSettings );
+  connect( job, &QgsMapRendererJob::finished, this, &QgsTerrainTextureGenerator::onRenderingFinished );
+  job->start();
+
+  JobData jobData;
+  jobData.jobId = ++mLastJobId;
+  jobData.job = job;
+  jobData.extent = extent;
+  jobData.debugText = debugText;
+
+  mJobs.insert( job, jobData );
+  //qDebug() << "added job: " << jobData.jobId << "  .... in queue: " << jobs.count();
+  return jobData.jobId;
+}
+
+void QgsTerrainTextureGenerator::cancelJob( int jobId )
+{
+  Q_FOREACH ( const JobData &jd, mJobs )
+  {
+    if ( jd.jobId == jobId )
+    {
+      //qDebug() << "canceling job " << jobId;
+      jd.job->cancelWithoutBlocking();
+      disconnect( jd.job, &QgsMapRendererJob::finished, this, &QgsTerrainTextureGenerator::onRenderingFinished );
+      jd.job->deleteLater();
```

Comment:
Is this safe? what if the job hasn't cancelled when it's deleted?

Reply:
That should be safe: job's destructor will wait until the job is fully cancelled